### PR TITLE
network:tc add support for qdisc handle

### PR DESCRIPTION
**Pull request by ssahani**
```
[Match]
Name=dummy99

[Network]
Address=192.168.3.11/24

[QDisc]
Parent=ingress
Handle=ffff

❯ sudo tc qdisc                                                                                                                                                                            ─╯

qdisc ingress ffff: dev dummy99 parent ffff:fff1 ---------------
```

### CODE REVIEW

File: src/network/tc/tc-util.c [hunk 1]
```diff
@@ -92,3 +93,25 @@ int tc_fill_ratespec_and_table(struct tc_ratespec *rate, uint32_t *rtab, uint32_
...
+                *h = m;
+                return 0;
+        }
+
+        m = strtoul(s, &p, 16);
+        if (p == s || m >= (1 << 16))
+                return -EINVAL;
+
+        m <<= 16;
+        if (*p != ':' && *p != 0)
```
Comment (yuwata) [inline src/network/tc/tc-util.c]: For qdisc, it is not necessary to add `:` at the end, as the minor number never specified. Let's simply takes major number without colon. (Of course, I know `tc` command takes colon. But it is not necessary to follow all syntax.)

Comment (ssahani) [reply src/network/tc/tc-util.c]: updated yes.

File: man/systemd.network.xml [hunk 2]
```diff
@@ -2263,6 +2263,14 @@
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+      <term><varname>Handle=</varname></term>
+      <listitem>
+        <para>Specifies the unique identifier is known as a handle and has two constituent members, a major number and a minor number.
+        or <literal>none</literal>. Defaults to unset.</para>
```
Comment (poettering) [inline man/systemd.network.xml]: please indicate what kind of value this expects, i.e. integer in which range

Comment (poettering) [reply man/systemd.network.xml]: also, specify this must be in hex

File: src/network/tc/tc-util.c [hunk 3]
```diff
@@ -92,3 +93,25 @@ int tc_fill_ratespec_and_table(struct tc_ratespec *rate, uint32_t *rtab, uint32_
...
+        uint32_t m;
+        char *p;
+
+        m = TC_H_UNSPEC;
+        if (streq(s, "none")) {
+                *h = m;
+                return 0;
+        }
+
+        m = strtoul(s, &p, 16);
```
Comment (poettering) [inline src/network/tc/tc-util.c]: strtoul returns "unsigned long". Which is not the same as "uint32_t". Please be more careful with the types. 

Only safe way to call strtoul() is by setting errno = 0 first, and checking if it is set afterwards.

As I understand this value must be in the 16bit range? if so, just use safe_atou16_full(), it takes care of all of this correctly.

Comment (ssahani) [reply src/network/tc/tc-util.c]: Sorry for late reply . I was travelling. It's a 32 bit number . But we can use safe_atou16_full() then assign this back .

**Comment by ssahani** [comment]
Updated. Thanks for the review @poettering

File: man/systemd.network.xml [hunk 4]
```diff
@@ -2277,6 +2277,14 @@
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+      <term><varname>Handle=</varname></term>
+      <listitem>
+        <para>Specifies the unique identifier is known as a handle and has two constituent members, a major number and a
+        minor number in hexadecimal digits or <literal>none</literal>. Defaults to unset.</para>
```
Comment (yuwata) [inline man/systemd.network.xml]: For qdisc, only major number is specified, isn't it?

File: src/network/tc/qdisc.c [hunk 5]
```diff
@@ -261,7 +262,9 @@ int config_parse_tc_qdiscs_parent(
 
         if (streq(rvalue, "root")) {
                 qdisc->parent = TC_H_ROOT;
-                qdisc->handle = TC_H_UNSPEC;
+
+                if (qdisc->handle == 0)
+                        qdisc->handle = TC_H_UNSPEC;
         } else if (streq(rvalue, "clsact")) {
                 qdisc->parent = TC_H_CLSACT;
                 qdisc->handle = TC_H_MAKE(TC_H_CLSACT, 0);
```
Comment (yuwata) [inline src/network/tc/qdisc.c:277]: This should be also
```
if (qdisc->handle == 0)
    qdisc->handle = TC_H_MAKE(TC_H_CLSACT, 0);
```

File: src/network/tc/tc-util.c [hunk 6]
```diff
@@ -92,3 +93,27 @@ int tc_fill_ratespec_and_table(struct tc_ratespec *rate, uint32_t *rtab, uint32_
...
+                return 0;
+        }
+
+        r = safe_atou16_full(s, 16, &n);
+        if (r < 0)
+                return r;
+
+        m = n;
+        if (m >= (1 << 16))
+                return -EINVAL;
```
Comment (yuwata) [inline src/network/tc/tc-util.c]: Please drop this check, as it is always negative.

**Comment by ssahani** [comment]
updated thanks for review @yuwata

**Comment by yuwata** [comment]
Rebased, and added tests.

**Comment by ssahani** [comment]
Ohh you just pushed :) thanks. I just finished it . LGTM

**Comment by yuwata** [comment]
Oh, sorry for duplicated work. Thank you for the review. Let's wait for the CIs.

**Pull request merged**